Makefile

```diff
@@ -51,57 +62,62 @@ OPENOCD := openocd
 DDD     := ddd
 GDB     := $(PREFIX)-gdb
 
-#application libraries
-include libs.mk
-
-CFLAGS ?= -Os -g -Wall -fno-common -c -mthumb \
-	  -mcpu=$(CPU_TYPE) -MD -std=gnu99
+# This file is dynamically created based on the libraries in the libs/ folder
+-include libs.mk
 
 INCLUDES += $(CPU_INCLUDES) $(BOARD_INCLUDES) $(LIB_INCLUDES) $(APP_INCLUDES)
+INCLUDES += -Icpu/common -Iboard/common
+
+CFLAGS ?= -Os -g -Wall -fno-common -c -mthumb
+CFLAGS += -mcpu=$(CPU_TYPE) -MD -std=gnu99
+#CFLAGS += -ffunction-sections -fdata-sections -flto
 CFLAGS += $(INCLUDES) $(CPU_DEFINES) $(BOARD_DEFINES) $(APP_DEFINES) $(CPU_FLAGS)
+
 ASFLAGS += -mcpu=$(CPU_TYPE) $(FPU) -g -Wa,--warn
 
 LIBS = -lnosys
-LIBS += $(addprefix -l,$(BASE_LIBS))
 
-LDFLAGS ?= --specs=nano.specs -lc -lgcc $(LIBS) -mcpu=$(CPU_TYPE) -g -gdwarf-2 \
-	-L. -Lcpu/common -L$(CPU_BASE) -T$(CPU_LINK_MEM) -Tlink_sections.ld \
-	-nostartfiles -Wl,--gc-sections -mthumb -mcpu=$(CPU_TYPE) \
-	-msoft-float -Wl,--Map=$(TARGET).map
+LDFLAGS ?= --specs=nano.specs -lc -lgcc $(LIBS) -mcpu=$(CPU_TYPE) -g -gdwarf-2 
+LDFLAGS += -L. -Lcpu/common -L$(CPU_BASE) -T$(CPU_LINK_MEM) -Tlink_sections.ld 
+LDFLAGS += -nostartfiles -Wl,--gc-sections -mthumb -mcpu=$(CPU_TYPE)
+#LDFLAGS += -ffunction-sections -fdata-sections -flto -Os
+LDFLAGS += -msoft-float -Wl,--Map=$(TARGET).map
 
 # Be silent per default, but 'make V=1' will show all compiler calls.
 ifneq ($(V),1)
-
 Q := @
 # Do not print "Entering directory ...".
 MAKEFLAGS += --no-print-directory
 endif
 
-# common objects
-OBJS += $(CPU_OBJS) $(BOARD_OBJS) $(APP_OBJS)
+# Object files
+APP_O_FILES = $(APP_C_FILES:.c=.o) $(APP_S_FILES:.s=.o)
```

Comment:
Good question.  There is a little bit of meta-make going on in this file (the libraries), so I had been wanting to expand as late as possible.  I can change to an expand-at-assignment and re-structure a little bit to only require this with the libraries.